drivers/sensor/hts221/hts221_trigger.c

```diff
@@ -15,154 +15,156 @@
 
 LOG_MODULE_DECLARE(HTS221, CONFIG_SENSOR_LOG_LEVEL);
 
-static inline void setup_drdy(struct hts221_data *drv_data,
-				  bool enable)
+static inline void setup_drdy(struct device *dev,
+			      bool enable)
 {
+	struct hts221_data *data = dev->driver_data;
+	const struct hts221_config *cfg = dev->config->config_info;
 	unsigned int flags = enable
 		? GPIO_INT_EDGE_TO_ACTIVE
 		: GPIO_INT_DISABLE;
 
-	gpio_pin_interrupt_configure(drv_data->gpio,
-				     DT_INST_0_ST_HTS221_DRDY_GPIOS_PIN,
-				     flags);
+	gpio_pin_interrupt_configure(data->drdy_dev, cfg->drdy_pin, flags);
 }
 
-static inline void handle_drdy(struct hts221_data *drv_data)
+static inline void handle_drdy(struct device *dev)
 {
-	setup_drdy(drv_data, false);
+	struct hts221_data *data = dev->driver_data;
+
+	setup_drdy(dev, false);
 
 #if defined(CONFIG_HTS221_TRIGGER_OWN_THREAD)
-	k_sem_give(&drv_data->gpio_sem);
+	k_sem_give(&data->drdy_sem);
 #elif defined(CONFIG_HTS221_TRIGGER_GLOBAL_THREAD)
-	k_work_submit(&drv_data->work);
+	k_work_submit(&data->work);
 #endif
 }
 
+static void process_drdy(struct device *dev)
+{
+	struct hts221_data *data = dev->driver_data;
+
+	if (data->data_ready_handler != NULL) {
+		data->data_ready_handler(dev, &data->data_ready_trigger);
+	}
+
+	if (data->data_ready_handler != NULL) {
```

Comment:
Is this correct? Why doing the same test twice?

Reply:
Because the user-provided handler could have reconfigured the GPIO to disable the interrupt.

Reply:
got it